EssentialApp/EssentialAppTests/CommentsUIIntegrationTests.swift

```
@@ -0,0 +1,199 @@
+		
+		loader.completeCommentsLoading(at: 0)
+		XCTAssertFalse(sut.isShowingLoadingIndicator, "Expecting no loading indicator once loading is completed successfully")
+	
+		sut.simulateUserInitiatedCommentsReload()
+		XCTAssertTrue(sut.isShowingLoadingIndicator, "Expecting a loading indicator once user initiated a reload")
+	
+		loader.completeCommentsLoadingWithError(at: 1)
+		XCTAssertFalse(sut.isShowingLoadingIndicator, "Expecting no loading indicator once user initiated loading is failed with error")
+	}
+	
+	func test_loadImageCommentsCompletion_rendersSuccessfullyLoadedComments() {
+		let comment0 = makeComment(message: "Comment 0", author: "Author 0")
+		let comment1 = makeComment(message: "Comment 1", author: "Author 1")
+		let comment2 = makeComment(message: "Comment 2", author: "Author 2")
+		let (sut, loader) = makeSUT()
+		
+		sut.loadViewIfNeeded()
+		assert(sut, isRendering: [])
+		
+		loader.completeCommentsLoading(with: [comment0], at: 0)
+		assert(sut, isRendering: [comment0])
+		
+		sut.simulateUserInitiatedCommentsReload()
+		loader.completeCommentsLoading(with: [comment0, comment1, comment2], at: 1)
+		assert(sut, isRendering: [comment0, comment1, comment2])
+	}
+	
+	func test_loadImageCommentsCompletion_doesNotAlteringCurrentRenderingStateOnError() {
+		let comment0 = makeComment()
+		let (sut, loader) = makeSUT()
+		
+		sut.loadViewIfNeeded()
+		loader.completeCommentsLoading(with: [comment0], at: 0)
+		assert(sut, isRendering: [comment0])
+		
+		sut.simulateUserInitiatedCommentsReload()
+		loader.completeCommentsLoadingWithError(at: 1)
+		assert(sut, isRendering: [comment0])
+	}
+	
+	func test_loadCommentsCompletion_dispatchesFromBackgroundToMainThread() {
+		let (sut, loader) = makeSUT()
+		
+		sut.loadViewIfNeeded()
+		
+		let exp = expectation(description: "Wait for loading completes")
+		DispatchQueue.global().async {
+			loader.completeCommentsLoading(with: [], at: 0)
+			exp.fulfill()
+		}
+		
+		wait(for: [exp], timeout: 1.0)
+	}
+	
```

Missing tests for:

- Error is shown when there's an error
- Error is hidden when the user initiates reload
- Error is hidden when the user taps on the error
- Request is canceled when the view controller (sut) is deallocated - You may need an explicit `autoreleasepool` to guarantee the `sut` is deallocated within the test method. For example: https://github.com/essentialdevelopercom/essential-feed-case-study/blob/master/EssentialApp/EssentialAppTests/CommentsUIIntegrationTests.swift#L137